### ISSUE TITLE
New Git Branch Validation Updates

### ISSUE THREAD

**Issue by JoelProminic** [prominic/Moonshine-IDE#751]
When I try to create a new Git branch, I get this popup:
![image](https://user-images.githubusercontent.com/24532731/98278503-5053fb00-1f5e-11eb-9b82-9c4585842843.png)

If the branch name is valid, this should continue without a prompt.

Normally, I'd like to see errors for this check appear in the form before submitting.  However, it seems like we are using a separate command to validate the branch name here, so this may not make sense:  https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/git/commands/CheckBranchNameAvailabilityCommand.as#L39

**Comment by rat-moonshine** [comment]
I see this is taking me more time than what I was expected earlier. 

While testing I see the earlier command to verify branch was not working correctly. The same results I also noticed when run on the command-line. The command always returns me the verifying branch name instead of any other data, be the branch exists or not remotely:

`git  check-ref-format --branch $branchName`

I change the command slightly, and this now properly returns me result to validate a branch existence remotely:

`git  show-ref --heads $branchName`

However, I turned into a rabbithole while applying the changes. For some reason, the creation of a git-command class always returns me the same instance that I created earlier even with the instantiation of `new` keyword, i.e. `new ClassName()`. This looks like taking a troll on me understanding why this happens. I shall commit my changes once I able to get an answer to this strange problem.

**Comment by rat-moonshine** [comment]
This issue is now ready for a test:

- New remote-branch-check command
- Revised handling of the above command return
- Showing no alert for a non-existing branch request and proceed to create the branch

**Comment by piotrzarzycki21** [comment]
@Aszusz please test this one, talk with @rat-moonshine about the details.

**Comment by JoelProminic** [comment]
I revisited this today on macOS:
![image](https://user-images.githubusercontent.com/24532731/101800711-2e0c4a80-3ad3-11eb-9714-00d47c8ce394.png)

The message should be something like this instead:

> Branch %branch-name% already exists.

However, this got me thinking about whether we want to distinguish between names that duplicate local branches, and names that duplicate remote branches only.   I'm also not sure if the current command can handle the remote-only branch detection.  This needs more testing.

**Comment by rat-moonshine** [comment]
For a reference, this is the current command that Moonshine executes to test branch-name availability:

`git show-ref --heads $branchName`

**Comment by Aszusz** [comment]
> However, this got me thinking about whether we want to distinguish between names that duplicate local branches, and names that duplicate remote branches only.

In my opinion we should distinguish between duplicating local branch and duplicating remote branch. This is because if I'm duplicating local branch half of the time my solution to this problem would be to delete existing branch and try again instead of thinking about different name.

> I'm also not sure if the current command can handle the remote-only branch detection. This needs more testing.

I am testing this today. It looks like current implementation can only detect local duplicates. Here's what happens when you try to duplicate remote only branch and push immediately:

![image](https://user-images.githubusercontent.com/7197014/101980514-3648bf00-3c66-11eb-9b1d-d74ba0747bdb.png)

I'll try to look for a git plumbing command that checks for remote branch existence.

**Comment by Aszusz** [comment]
OK, looking for a duplicate against remote branches is a little bit more complicated, but nothing too crazy.

1. First we need to know which remote we'd like to check. Therefore we should obtain a list of all our remotes with `git remote -v`
![image](https://user-images.githubusercontent.com/7197014/101981003-e4a23380-3c69-11eb-82f0-e81fe71e6390.png)

2. Then we should parse the output and get second column.

3. As you can see this list contains pairs of fetch+push addresses which 99% of the time are the same. That's why we should filters this list for distinct addresses.

4. Then, for each distinct address we should run a branch check `git ls-remote $repo-name --heads $branch-name`:
![image](https://user-images.githubusercontent.com/7197014/101981116-dacd0000-3c6a-11eb-9eaf-4b77135847f8.png)
This is an equivalent of the command @rat-moonshine used for checking local duplicates.

**Comment by rat-moonshine** [comment]
> First we need to know which remote we'd like to check. Therefore we should obtain a list of all our remotes with git remote -v

Currently, we are retrieving an opened project's repository address in this way: https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/git/commands/GetRemoteURLCommand.as#L40. 

Could it be equivalent to what `git remote -v` does to retrieve the repository address? If so then we already have the URL parsed to the opened project's git-objects.

> Then, for each distinct address we should run a branch check git ls-remote $repo-name --heads $branch-name:
> This is an equivalent of the command @rat-moonshine used for checking local duplicates.

The above command is supposed to check the remote brances. Still, we are not clear how to distinguish the remote and local branch name-duplicates visually from the 'Create Branch' popup, as there is no option to determine currently which duplicate path check is performed - local or remote.

**Comment by Aszusz** [comment]
> Could it be equivalent to what git remote -v does to retrieve the repository address?

In a lot of cases the result will be the same, however I can spot a few differences:
- As far as I understand the current code, it gives you one address per repository. There are cases when people work with multiple upstream/downstream remotes and need a list of addresses. Is this relevant for Moonshine users -- I don't know.
- Your regex checks for `http.*` -- when you clone with ssh your remote looks like this: `git@github.com:prominic/Moonshine-IDE.git`

> The above command is supposed to check the remote brances.

Not sure to which of the commands you're referring, so I'll sum it up:
- `git show-ref --heads $branchName` checks local branches **only**
- `git ls-remote $repoName --heads $branchName` checks remote branches **only**

> Still, we are not clear how to distinguish the remote and local branch name-duplicates visually from the 'Create Branch' popup, as there is no option to determine currently which duplicate path check is performed - local or remote.

I think to do this you need to somehow preserve the knowledge whether you used `git show-ref` or `git ls-remote`.

**Comment by JoelProminic** [comment]
I was able to setup a project with multiple origins like this:

    git clone </path/to/local/repository/.git> <New-Project-Name>
    cd <New-Project-Name>
    git remote add github https://github.com/...
    git pull github 


I see output like this:

    $ git remote -v       
    github	https://github.com/JoelProminic/Test-Private-Repository.git (fetch)
    github	https://github.com/JoelProminic/Test-Private-Repository.git (push)
    origin	/Users/-username-/Documents/Flex/temp_workspace/Test-Private-Repository.git/.git (fetch)
    origin	/Users/-username-/Documents/Flex/temp_workspace/Test-Private-Repository.git/.git (push)
    $ git branch --remote 
      github/TestBranch
      github/TestRemote
      github/main
      github/test-11-25
      github/test-12-04
      github/test-remote-12-04
      origin/HEAD -> origin/main
      origin/TestBranch
      origin/TestRemote
      origin/main
      origin/test-11-25
      origin/test-12-04
      origin/test-remote-12-04

Could we use the `git branch` output to check for duplicates, or is there an advantage to testing with `git show-ref` and `git ls-remote`.

Either way, we can return updated error messages.  For a local branch:

> Local branch '<branchname>' already exists.

Remote.  I think it is adequate to show the first remote where this is found, but we could list all matching remotes if desired:

> Branch '<branchname>' already exists on remote '<remote>'.

Note that we don't have multiple origin support for Push or Pull currently, so this may be overkill for our current use cases.

**Comment by Aszusz** [comment]
> Could we use the git branch output to check for duplicates, or is there an advantage to testing with git show-ref and git ls-remote.

There are advantages:
- `git branch --remote` only shows you locally cached refs of the remote. To be accurate you need to do `git fetch` first.
- if you already know the name of the branch using `git ls-remote $repoName --heads $branchName` is easier because it gives you some text output if the branch exists (true), or no text output if the branch doesn't exist (false) -- no need for additional parsing branch names.

Why would you prefer to use `git branch --remote` for this? Is this because making a list of remotes is hard for you? 
Would it help if I came up with one-line command combining remote iteration with ls-remote?
I consider this: `git remote | xargs -I % sh -c "git ls-remote % --heads branch-name"`, for now I tested this on Windows.


> Note that we don't have multiple origin support for Push or Pull currently, so this may be overkill for our current use cases.

Fine by me, if we assume there is only one remote we can also assume its name is origin (should be 99% of the time) and simplify the command to: `git ls-remote origin --heads $branchName` (git will replace origin with the correct address by itself).

**Comment by rat-moonshine** [comment]
Since this going to be more work than the initial branch-name-check that we imagined in the original issue, I'm moving the new change-implementation to a separate branch - https://github.com/prominic/Moonshine-IDE/tree/features/issue_751_new_branch_validation.

**Comment by JoelProminic** [comment]
> OK, looking for a duplicate against remote branches is a little bit more complicated, but nothing too crazy.
> 
>     1. First we need to know which remote we'd like to check. Therefore we should obtain a list of all our remotes with `git remote -v`
>        ![image](https://user-images.githubusercontent.com/7197014/101981003-e4a23380-3c69-11eb-82f0-e81fe71e6390.png)
> 
>     2. Then we should parse the output and get second column.

@Aszusz, is it necessary to use the full URL for the remotes?   Using the names seemed to work fine in my quick test:

    $ git remote
    github
    origin
    $ git branch --all
    * main
      remotes/github/TestBranch
      remotes/github/TestRemote
      remotes/github/main
      remotes/github/test-11-25
      remotes/github/test-12-04
      remotes/github/test-remote-12-04
      remotes/origin/HEAD -> origin/main
      remotes/origin/TestBranch
      remotes/origin/TestRemote
      remotes/origin/main
      remotes/origin/test-11-25
      remotes/origin/test-12-04
      remotes/origin/test-remote-12-04
    $ git ls-remote github --heads TestBranch
    2910051c7bea28d1d834bc321d0fc81c06335cbd	refs/heads/TestBranch
    $ git ls-remote origin --heads TestBranch
    2910051c7bea28d1d834bc321d0fc81c06335cbd	refs/heads/TestBranch
    2910051c7bea28d1d834bc321d0fc81c06335cbd	refs/remotes/origin/TestBranch

**Comment by Aszusz** [comment]
@JoelProminic 
> is it necessary to use the full URL for the remotes?
No, using just the names is absolutely fine.

**Comment by rat-moonshine** [comment]
> @JoelProminic
> 
> > is it necessary to use the full URL for the remotes?
> 
> No, using just the names is absolutely fine.

If it is so, then I believe we don't need the `git remote -v` and parsing its origins [separately](https://github.com/prominic/Moonshine-IDE/issues/751#issuecomment-743734258) (?)

Instead, running these two commands would be sufficient to check between local and remote branches, I guess.. 

Local: `git show-ref --heads $branchName`
Remote: `git ls-remote github --heads $branchName` ('github' or 'origin')

**Comment by rat-moonshine** [comment]
Considering the [above](https://github.com/prominic/Moonshine-IDE/issues/751#issuecomment-745776099) assumption is correct, I have updated the name validation test codes and this seems to be working fine. The changes also merged into the 'master'.

Branch Name | Exists in Local | Exists in Remote | Results
--------------- | ------------------ | ---------------- | -------------
TestLocalBranch | Yes | No | Message shown exists at local
TestRemoteBranch (created from another system while testing) | No | Yes | Message shown exists at remote
TestOldBranch | Yes | Yes | Message shown exists at remote and local
TestNewBranch | No | No | Moonshine creates a new branch

**Comment by Aszusz** [comment]
> If it is so, then I believe we don't need the git remote -v and parsing its origins separately (?)

No, however, passing output of plain `git remote` to `git ls-remote` would be nice.

> Considering the above assumption is correct, I have updated the name validation test codes and this seems to be working fine. The changes also merged into the 'master'.

I tested your changes. When the remote is named 'origin' everything works as expected.

But when I renamed my remote from 'origin' to 'custom' I got this error:
`: fatal: 'origin' does not appear to be a git repository`
and I can no longer create **any** branch. I think if we decide to hardcode 'origin' into the command we should at least allow the user to create (possibly duplicate) branch when the command fails with an error.

**Comment by rat-moonshine** [comment]
> But when I renamed my remote from 'origin' to 'custom' I got this error:
> : fatal: 'origin' does not appear to be a git repository
> and I can no longer create any branch. I think if we decide to hardcode 'origin' into the command we should at least allow the user to create (possibly duplicate) branch when the command fails with an error.

Is the above problem can solve if we return to your earlier solution by parsing available origins and by passing the URL instead of 'origin' as [here](https://github.com/prominic/Moonshine-IDE/issues/751#issuecomment-743734258)?

**Comment by Aszusz** [comment]
> Is the above problem can be resolved if we return to your earlier solution by parsing available origins and by passing the URL instead of 'origin' as here?

Yes. But we can make one significant simplification: as @JoelProminic noticed, we don't need URLs, we can use names, as long as they are actual names, coming out of `git remote`.

Let me modify slightly my previous solution:
1. First we need to know which remote we'd like to check. Therefore we should obtain a list of all our remotes with `git remote`
![image](https://user-images.githubusercontent.com/7197014/102342318-5e535d80-3f99-11eb-92a2-f0bd30b1f257.png)
2. Then we should use the whole output of this command -- no need for parsing.
3. For each remote name we run a branch check `git ls-remote $remoteName --heads $branchName`

And as I suggested before, if doing this from ActionScript seems hard, you could run a single command that combines steps 1, 2 and 3. For example:
`git remote | xargs -I % sh -c "git ls-remote % --heads $branchName"`
I'm not saying you should use this exact line, just that it's possible to do in one line if you want.

**Comment by rat-moonshine** [comment]
Yeah, I realized that I have overlooked the usage of `git remote` from Joel's example commands, too. Which made me falsely think that using static 'origin' or 'github' keyword is by some standard process.

The suggested steps from [1-3](https://github.com/prominic/Moonshine-IDE/issues/751#issuecomment-746150629) can be implemented easily, I guess, @Aszusz ; I shall follow this path.

**Comment by rat-moonshine** [comment]
On a separate question, if the user has 5 origins, and 3 out of 5 has the branch name pre-exists, and also exists at local, how the message we would like to expect?

- If 3 out of 5 Origins has the branch pre-exists, shall we require to list every origin that has the branch name pre-exists?
- Is the branch creation be permitted if 3 out of 5 origins has the branch pre-exists?

**Comment by Aszusz** [comment]
> If 3 out of 5 Origins has the branch pre-exists, shall we require to list every origin that has the branch name pre-exists?

In my opinion: no. I'd say the message can still be "exists at remote" and when the user goes to switch to branch -> remote he/she can check the remote name. @JoelProminic said to include first remote name and that's also fine for me.

> Is the branch creation be permitted if 3 out of 5 origins has the branch pre-exists?

In my opinion: no. If you are working with upstream/downstream remotes you usually want to bring them into sync in the end. So this is precisely the case when you want to know you need to pull a branch from remote instead of creating it once again.

**Comment by rat-moonshine** [comment]
I have updated this issue with multi-origin tests as returned by `git remote`. This [matrix](https://github.com/prominic/Moonshine-IDE/issues/751#issuecomment-745957758) is still valid following these changes. 

Please, check.

**Comment by JoelProminic** [comment]
Here are my testing results:

Test | Expected | Actual
-----|----------|-------
No duplicate | Branch created | Confirmed
Local duplicate only | Report error for local | "Local branch already exists;"
Remote duplicate only, single remote | Report error for remote | "Branch exists on remote;"
Remote duplicate only on first remote | Report error for first remote | "Branch exists on remote;"
Remote duplicate only on second remote | Report error for second remote | "Branch exists on remote;"
Remote duplicate only on both remotes | Report error for first remote | "Branch exists on remote"
Duplicate in all locations | Report error for local | "Local branch already exists; Branch exists on remote;"


All of the duplicates were detected, but I would like some tweaks to the messages:
- Include remote name in remote message.  We don't need to include the branch name, since it is displayed in the input field
- Use a period instead of a semicolon at the end of the messages
- Only display the first error, as described in my test matrix.  This will reduce the risk of long messages. 
- Add "already" to remote message to match local message
- Should be "in remote" rather than "on remote"

Here are the updated messages:

Local message:

> Local branch already exists.

Remote message:

> Branch already exists in remote '%remote_name%'.

**Comment by rat-moonshine** [comment]
Name validation messages updates with additional information according to the above suggestions. Please, check.

**Comment by JoelProminic** [comment]
I ran through the test cases, and the updated messages look good.

Test | Expected | Actual
-----|----------|-------
No duplicate | Branch created | 
Local duplicate only | Report error for local | "Local branch already exists"
Remote duplicate only, single remote | Report error for remote | "Branch already exists in remote: origin."
Remote duplicate only on first remote | Report error for first remote | "Branch already exists in remote: github."
Remote duplicate only on second remote | Report error for second remote | "Branch already exists in remote: origin."
Remote duplicate only on both remotes | Report error for first remote | "Branch already exists in remote: github."
Duplicate in all locations | Report error for local | "Local branch already exists"


However, I noticed that I couldn't switch to a branch on the additional remote I added:

    : Trying to switch branch...
    : Note: switching to 'github/TestRemote'.
    : You are in 'detached HEAD' state. You can look around, make experimental
    : changes and commit them, and you can discard any commits you make in this
    : state without impacting any branches by switching back to a branch.
    : If you want to create a new branch to retain commits you create, you may
    : do so (now or later) by using -c with the switch command. Example:
    :   git switch -c <new-branch-name>
    : Or undo this operation with:
    :   git switch -
    : Turn off this advice by setting config variable advice.detachedHead to false
    : M	obj/TestPrivateRepositoryConfig.xml

I get stuck in the "detached head" state.  I see some options to work with this [here](https://stackoverflow.com/questions/1783405/how-do-i-check-out-a-remote-git-branch), but it looks like they create a new local branch to match the remote.

@Aszusz, is there a simple way to support this case?  If we need to add a lot more complexity to the application to support multiple origins, we may want to save it for a future update.

Otherwise, this issue looks good.

**Comment by Aszusz** [comment]
> is there a simple way to support this case? If we need to add a lot more complexity to the application to support multiple origins, we may want to save it for a future update.

I think supporting multiple origins is something we can save for the future. In Git a local branch can track only one remote branch. If you want to sync a branch between two repositories you usually set tracking to repository named origin and manually pull from repository named upstream. So we would need a couple of changes to push/pull, etc.

My only concern for now is leaving the user in a detached head state. I'd like to avoid this if possible.
@rat-moonshine Could you tell me what parameters are we using for `git checkout` command? If none, then how the local tracking branch is created?

**Comment by rat-moonshine** [comment]
Looking through the _commands_ we're using, I see the following areas using `checkout`:

1. [During the creation of a new branch](https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/git/commands/CreateCheckoutNewBranchCommand.as#L43)
2. [Branch change command](https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/git/commands/GitChangeBranchToCommand.as#L40)
3. [Repository checkout command](https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/git/commands/GitCheckoutCommand.as#L42)
4. [Revert/restore command](https://github.com/prominic/Moonshine-IDE/blob/master/ide/MoonshineDESKTOPevolved/src/actionScripts/plugins/git/commands/RevertCommand.as#L50)

Probably 2 from the above where your interest is.

**Comment by JoelProminic** [comment]
It should be fine to merge this code to master so that we can have the updates for single-origin projects.  We will create a separate issue for switching branches on multiple-origin projects.

**Comment by rat-moonshine** [comment]
> 
> 
> It should be fine to merge this code to master so that we can have the updates for single-origin projects. We will create a separate issue for switching branches on multiple-origin projects.

It seem we were always fixing this on master only. What do we left before we can close this issue (?)

**Comment by JoelProminic** [comment]
I confirmed that I can see the updates in master.   It looks like the remaining concerns were moved to #769, so I'm closing this issue.
